### PR TITLE
Upgrade coffeescript

### PR DESCRIPTION
**Pull request by johnjeng**
Updates to Coffeescript 2, keeping the `.cjsx` extension and not transitioning to `import`/`export` syntax. 

# Tests
Util tests work while hub tests fail to import Coffeescript at all for some reason (ie. the tests don't run)
- [x] smc-util tests
- [ ] smc-hub tests
- [x] jupyter tests

### CODE REVIEW

**Comment by johnjeng** [comment]
@DrXyzzy Can you test this?

**Comment by DrXyzzy** [comment]
yes, today

**Comment by williamstein** [comment]
I can't build in Kubernetes prod due to:
```
 
npm WARN lifecycle smc-webapp@1.0.0~postinstall: cannot run in wd %s %s (wd=%s) smc-webapp@1.0.0 babel --presets=env  node_modules/prom-client -d node_modules/prom-client-js /c
ocalc/src/smc-webapp
npm WARN react-addons-perf@15.4.2 requires a peer of react-dom@^15.4.2 but none is installed. You must install peer dependencies yourself.
npm WARN The package coffeescript is included as both a dev and production dependency.
npm WARN The package coffeelint is included as both a dev and production dependency.
npm WARN The package expect is included as both a dev and production dependency.
npm WARN The package mocha is included as both a dev and production dependency.
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.1.3 (node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.1.3: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})
 
added 726 packages in 24.722s
npm WARN cocalc-jupyter@1.0.0 No repository field. 
npm WARN cocalc-jupyter@1.0.0 license should be a valid SPDX license expression
 
added 77 packages in 2.866s
Error: Using `babel-preset-react-app` requires that you specify `NODE_ENV` or `BABEL_ENV` environment variables. Valid values are "development", "test", and "production". Inste
ad, received: undefined. (While processing preset: "/cocalc/src/node_modules/babel-preset-react-app/index.js")
    at Object.<anonymous> (/cocalc/src/node_modules/babel-preset-react-app/index.js:47:9)
    at Module._compile (module.js:643:30)
    at Object.Module._extensions..js (module.js:654:10)
    at Module.load (module.js:556:32)
    at tryModuleLoad (module.js:499:12)
    at Function.Module._load (module.js:491:3)
    at Module.require (module.js:587:17)
    at require (internal/module.js:11:18)
    at /cocalc/src/smc-webapp/node_modules/babel-core/lib/transformation/file/options/option-manager.js:296:17
    at Array.map (<anonymous>)
umask 022; cd examples && make
```

**Comment by williamstein** [comment]
however, **Projects** on ws1 are now running with cs2.  The frontend can't until the above is fixed.

**Comment by williamstein** [comment]
I'm making further attempts to build this on ws1.  If I succeed, I'll post again, and then it'll be time to test -- by simply using ws1 (after a browser refresh!)

**Comment by williamstein** [comment]
OK, right now https://ws1.cocalc.com/ is running [@j3's coffeescript2 branch right now](https://github.com/sagemathinc/cocalc/pull/2637), everywhere -- hubs, manage, frontend, and project.   PLEASE TEST!

**Comment by DrXyzzy** [comment]
tests in ws1:

## sagews pytest
```
cd ~/cocalc/src/smc_sagews/smc_sagews/tests
python -m pytest
= 141 passed, 5 skipped in 359.97 seconds =
```

## mocha
```
cd ~/cocalc/src
npm run test
... ok for smc-util and smc-util-node
TEST POSTGRES
/home/user/cocalc/src/smc-hub/node_modules/coffeescript/lib/coffeescript/index.js:143
        throw new Error(`Use CoffeeScript.register() or require the coffeescript/register module to require ${ext} files.`);
        ^
Error: Use CoffeeScript.register() or require the coffeescript/register module to require .coffee files.
```
Same `Use CoffeeScript.register()` error with
```
cd ~/cocalc/src/smc-project
npm run test
```

**Comment by DrXyzzy** [comment]
@johnjeng - Nice. I had to make a couple minor changes to get mocha smc-project tests to pass in my project in ws1, with your latest commits just now to the upgrade-coffeescript branch. After the changes below, I see
```
cd ~/cocalc/src/dev-project
npm run test
...
TEST PROJECT
...
  4 passing (2s)

TEST JUPYTER SERVER
  77 passing (2m)
```

Changes:

1. Add `*.coffee` to test path in `src/smc-project/packages.json`:
```
diff --git a/src/smc-project/package.json b/src/smc-project/package.json
index d014cfe..3fd3329 100644
--- a/src/smc-project/package.json
+++ b/src/smc-project/package.json
@@ -34,7 +34,7 @@
     "coffeelint": "^2.0.7"
   },
   "scripts": {
-    "test": "echo 'TEST PROJECT'; node_modules/.bin/mocha --reporter progress test && echo 'TEST JUPYTER SERVER'; cd jupyter && ../node_modules/.bin/mocha
 --reporter  progress  test"
+    "test": "echo 'TEST PROJECT'; node_modules/.bin/mocha --reporter progress test/*.coffee && echo 'TEST JUPYTER SERVER'; cd jupyter && ../node_modules/.
bin/mocha --reporter  progress  test/*.coffee"
   },
   "author": "SageMath, Inc.",
```

2. Bump timeouts for some of the slower kernels in `src/smc-project/jupyter/test/supported-kernels.coffee`. Often tests timeout on first run but pass if the tests are re-run immediately:
```
diff --git a/src/smc-project/jupyter/test/supported-kernels.coffee b/src/smc-project/jupyter/test/supported-kernels.coffee
index 27e9d56..1908805 100644
--- a/src/smc-project/jupyter/test/supported-kernels.coffee
+++ b/src/smc-project/jupyter/test/supported-kernels.coffee
@@ -132,7 +132,7 @@ describe 'test the octave kernel --', ->
 
 
 describe 'test the julia kernel --', ->
-    @timeout(20000)
+    @timeout(80000)
     kernel = undefined
     it 'evaluate 4/3', (done) ->
         kernel = common.kernel('julia')
@@ -151,7 +151,7 @@ describe 'test the julia kernel --', ->
 
 
 describe 'test the system-wide R kernel --', ->
-    @timeout(60000)
+    @timeout(80000)
     kernel = undefined
     it 'evaluate sd(c(1,2,5))', (done) ->
         kernel = common.kernel('ir')
@@ -169,7 +169,7 @@ describe 'test the system-wide R kernel --', ->
         kernel.close()
 
 describe 'test the sage R kernel --', ->
-    @timeout(60000)
+    @timeout(80000)
     kernel = undefined
     it 'evaluate sd(c(1,2,5))', (done) ->
         kernel = common.kernel('ir-sage')
@@ -187,7 +187,7 @@ describe 'test the sage R kernel --', ->
         kernel.close()

 describe 'test the scala kernel --', ->
-    @timeout(30000)
+    @timeout(80000)
     kernel = undefined
     it 'matchTest', (done) ->
         kernel = common.kernel('scala211')
```

**Comment by DrXyzzy** [comment]
Notes after review of mocha tests in ws1 in upgrade-coffeescript branch:

1. Tests in this directory fail summarily.
```
cd cocalc/src/smc-webapp/jupyter
../node_modules/.bin/mocha test/*.coffee
... 
coffee-cache: No coffee-script package found.
/home/user/cocalc/src/smc-webapp/jupyter/test/setup.coffee:12
  require('coffee-cache').setCacheDir(`${process.env.HOME}/.coffee/cache`);
                          ^

TypeError: require(...).setCacheDir is not a function
```

2. The following test also fails in production, so failing in ws1 cs2 branch might not be due to cs2:
```
cd cocalc/src/smc-webapp
./node_modules/.bin/mocha test/smc-react-test.coffee
... last test fails: it 'Recomputes once on one input change' ...
```

**Comment by williamstein** [comment]
Current blocker @DrXyzzy found, which is that 

https://ws1.cocalc.com/projects/008338c5-7817-478d-b0e2-f5d94261400f/log?session=default

causes this traceback:

screenshot 2018-02-10 at 2 59 27 pm

File: src/smc-hub/postgres.coffee [hunk 1]
```diff
@@ -1,28 +1,30 @@
...
 
-exports.PUBLIC_PROJECT_COLUMNS = ['project_id',  'last_edited', 'title', 'description', 'deleted',  'created']
-exports.PROJECT_COLUMNS = ['users'].concat(exports.PUBLIC_PROJECT_COLUMNS)
-
+exports.PUBLIC_PROJECT_COLUMNS = base.PUBLIC_PROJECT_COLUMNS
+exports.PROJECT_COLUMNS        = base.PROJECT_COLUMNS
 
 # Add further functionality to PostgreSQL class -- must be at the bottom of this file.
-# Each of the following calls extends the PostgreSQL class with further important functionality.
+# Each of the following calls composes the PostgreSQL class with further important functionality.
```
Comment (williamstein) [inline src/smc-hub/postgres.coffee]: got to say that changing extends to composes is inconsistent with `extend_PostgreSQL` below...

**Pull request merged**